Framework/Indexing/inc/MantidIndexing/IndexSet.h

```diff
@@ -0,0 +1,152 @@
+  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+  File change history is stored at: <https://github.com/mantidproject/mantid>
+  Code Documentation is available at: <http://doxygen.mantidproject.org>
+*/
+template <class T> class IndexSet {
+public:
+  class const_iterator
+      : public boost::iterator_facade<const_iterator, size_t,
+                                      boost::random_access_traversal_tag,
+                                      size_t> {
+  public:
+    const_iterator(const IndexSet &indexSet, const size_t index)
+        : m_indexSet(indexSet), m_index(index) {}
+
+  private:
+    friend class boost::iterator_core_access;
+
+    void increment() {
+      if (m_index < m_indexSet.size())
+        ++m_index;
+    }
+
+    bool equal(const const_iterator &other) const {
+      return (&m_indexSet == &other.m_indexSet) && (m_index == other.m_index);
+    }
+
+    size_t dereference() const { return m_indexSet[m_index]; }
+
+    void decrement() {
+      if (m_index > 0)
+        --m_index;
+    }
+    void advance(int64_t delta) {
+      m_index = delta < 0 ? std::max(static_cast<int64_t>(0),
+                                     static_cast<int64_t>(m_index) + delta)
+                          : std::min(m_indexSet.size(),
+                                     m_index + static_cast<size_t>(delta));
+    }
+    int64_t distance_to(const const_iterator &other) const {
+      return static_cast<int64_t>(other.m_index) -
+             static_cast<int64_t>(m_index);
+    }
+
+    const IndexSet &m_indexSet;
+    size_t m_index;
+  };
+
+  const_iterator begin() const { return const_iterator(*this, 0); }
+  const_iterator end() const { return const_iterator(*this, size()); }
+
+  IndexSet(size_t fullRange);
+  IndexSet(int64_t min, int64_t max, size_t fullRange);
+  IndexSet(const std::vector<size_t> &indices, size_t fullRange);
+
+  /// Returns the size of the set.
+  size_t size() const { return m_size; }
+
+  /// Returns the element at given index (range 0...size()-1).
+  size_t operator[](size_t index) const {
+    // This is accessed frequently in loops and thus inlined.
+    if (m_isRange)
+      return m_min + index;
+    return m_indices[index];
+  }
+
+protected:
+  ~IndexSet() = default;
+
+private:
+  bool m_isRange = true;
+  // Default here to avoid uninitialized warning for m_isRange = false.
+  size_t m_min = 0;
+  size_t m_size;
+  std::vector<size_t> m_indices;
+};
+
+/// Constructor for a set covering the full range from 0 to fullRange-1.
+template <class T>
+IndexSet<T>::IndexSet(size_t fullRange)
+    : m_size(fullRange) {}
+
+/// Constructor for a set covering the range from min to max. Range is verified
+/// at construction time.
+template <class T>
+IndexSet<T>::IndexSet(int64_t min, int64_t max, size_t fullRange) {
+  if (min < 0 || min > max)
+    throw std::logic_error("IndexSet: specified min or max values are invalid");
+  if (max >= static_cast<int64_t>(fullRange))
```

Comment:
Just for validation, you would pass the workspace size and `max` is checked automatically. You can of course still break things by passing something invalid to `fullRange`, but if the range is correct, it is guaranteed that any index in the set will be a valid workspace index. I agree that is not strictly necessary.